src/consensus/aft/raft.h

```diff
@@ -1419,31 +1419,22 @@ namespace aft
         }
       }
 
-      // Update next and match for the responding node.
-      auto& match_idx = node->second.match_idx;
+      // Update next or match for the responding node.
       if (r.success == AppendEntriesResponseType::FAIL)
       {
-        const auto this_match =
-          find_highest_possible_match({r.term, r.last_log_idx});
-        if (match_idx == 0)
-        {
-          match_idx = this_match;
-        }
-        else
-        {
-          match_idx = std::min(match_idx, this_match);
-        }
         // Failed due to log inconsistency. Reset sent_idx, and try again soon.
         RAFT_DEBUG_FMT(
           "Recv append entries response to {} from {}: failed",
           state->node_id,
           from);
-        node->second.sent_idx = node->second.match_idx;
+        const auto this_match =
+          find_highest_possible_match({r.term, r.last_log_idx});
+        node->second.sent_idx = std::min(this_match, node->second.sent_idx);
         return;
       }
       else
       {
-        match_idx = std::min(r.last_log_idx, state->last_idx);
+        node->second.match_idx = std::min(r.last_log_idx, state->last_idx);
```

Comment:
Taking the `min` seems to still allow `match_idx` to go backwards if `r.last_log_idx` happens to be less than `state->last_idx`.  The spec had to be changed to take the `max` in the (unreviewed) PR #5324 to satisfy the new `MonotonicMatchIndexProp` property:

https://github.com/microsoft/CCF/pull/5324/commits/1454326be5a9f80822b892b31381d4c5bc4d3ebf

Reply:
+1, we probably want to log debug if/when this happens.

Reply:
>but I believe that's never possible due to the ordered handling of messages (guaranteed by the node-to-node channel) and the preconditions within this function.

Agreed, but it's clearer to make it explicit here and in the spec. We can add a comment that points to the guarantee coming from the node to node channel anyway.